test/simple_ng/simple_ng.c

```diff
@@ -0,0 +1,318 @@
+    }
+    printf("done\n");
+
+    printf("Finding autoconfig slaves... ");
+    if (ecx_config_init(context, FALSE) <= 0) {
+        printf("no slaves found\n");
+        return FALSE;
+    }
+    printf("%d slaves found\n", fieldbus->slavecount);
+
+    printf("Sequential mapping of I/O... ");
+    ecx_config_map_group(context, fieldbus->map, fieldbus->group);
+    printf("mapped %dO+%dI bytes from %d segments",
+           grp->Obytes, grp->Ibytes, grp->nsegments);
+    if (grp->nsegments > 1) {
+        /* Show how slaves are distrubuted */
+        for (i = 0; i < grp->nsegments; ++i) {
+            printf("%s%d", i == 0 ? " (" : "+", grp->IOsegment[i]);
+        }
+        printf(" slaves)");
+    }
+    printf("\n");
+
+    printf("Configuring distributed clock... ");
+    ecx_configdc(context);
+    printf("done\n");
+
+    printf("Waiting for all slaves in safe operational... ");
+    ecx_statecheck(context, 0, EC_STATE_SAFE_OP, EC_TIMEOUTSTATE * 4);
+    printf("done\n");
+
+    printf("Send a roundtrip to make outputs in slaves happy... ");
+    fieldbus_roundtrip(fieldbus);
+    printf("done\n");
+
+    printf("Setting operational state..");
+    /* Act on slave 0 (a virtual slave used for broadcasting) */
+    slave = fieldbus->slavelist;
+    slave->state = EC_STATE_OPERATIONAL;
+    ecx_writestate(context, 0);
+    /* Poll the result ten times before giving up */
+    for (i = 0; i < 10; ++i) {
+        printf(".");
+        fieldbus_roundtrip(fieldbus);
+        ecx_statecheck(context, 0, EC_STATE_OPERATIONAL, EC_TIMEOUTSTATE / 10);
+        if (slave->state == EC_STATE_OPERATIONAL) {
+            printf(" all slaves are now operational\n");
+            return TRUE;
+        }
+    }
+
+    printf(" failed,");
+    ecx_readstate(context);
+    for (i = 1; i <= fieldbus->slavecount; ++i) {
+        slave = fieldbus->slavelist + i;
+        if (slave->state != EC_STATE_OPERATIONAL) {
+            printf(" slave %d is 0x%04X (AL-status=0x%04X %s)",
+                 i, slave->state, slave->ALstatuscode,
+                 ec_ALstatuscode2string(slave->ALstatuscode));
+        }
+    }
+    printf("\n");
+
+    return FALSE;
+}
+
+static void
+fieldbus_stop(Fieldbus *fieldbus)
+{
+    ecx_contextt *context;
+    ec_slavet *slave;
+
+    context = &fieldbus->context;
+    /* Act on slave 0 (a virtual slave used for broadcasting) */
+    slave = fieldbus->slavelist;
+
+    printf("Requesting init state on all slaves... ");
+    slave->state = EC_STATE_INIT;
+    ecx_writestate(context, 0);
+    printf("done\n");
+
+    printf("Close socket... ");
+    ecx_close(context);
+    printf("done\n");
+}
+
+static boolean
+fieldbus_dump(Fieldbus *fieldbus)
+{
+    ec_groupt *grp;
+    uint32 n;
+    int wkc, expected_wkc;
+
+    grp = fieldbus->grouplist + fieldbus->group;
+
+    wkc = fieldbus_roundtrip(fieldbus);
+    expected_wkc = grp->outputsWKC * 2 + grp->inputsWKC;
+    printf("%d usec  WKC %d", fieldbus->roundtrip_time, wkc);
```

Comment:
Add a fixed print format, otehrwise it becomes a big blur on the output.
Ex below is nicer to look at.
    printf("%3d usec  WKC %d", fieldbus->roundtrip_time, wkc);

Reply:
This would limit the time to 1 ms. On some test I did in harsh conditions, I've seen roundtrip times **much** longer than that. For this reason I'd rather prefer to leave the length free.
Alternatively, considering the other fields are almost fixed-size, I can move it at the end of the line, so it can grow as it wishes without messing up everything.

Reply:
Sure, as long as it is possible to see what’s is going on

Reply:
Moving that field to the end was not feasible because while the roundtrip time is always present, I do not output some of the other fields if the wkc does not match. I just went to the padding root with more loose limits, i.e. 6 digits for the roundtrip time and 4 digits for the iteration counter.